modules/core/src/main/scala/org/tessellation/infrastructure/snapshot/GlobalSnapshotLoader.scala

```diff
@@ -0,0 +1,49 @@
+package org.tessellation.infrastructure.snapshot
+
+import cats.data.OptionT
+import cats.effect.kernel.Async
+import cats.syntax.either._
+import cats.syntax.functor._
+
+import org.tessellation.kryo.KryoSerializer
+import org.tessellation.schema.{GlobalIncrementalSnapshot, GlobalSnapshot}
+import org.tessellation.sdk.infrastructure.snapshot.storage.SnapshotLocalFileSystemStorage
+import org.tessellation.security.hash.Hash
+import org.tessellation.security.signature.Signed
+
+import fs2.io.file.Path
+
+trait GlobalSnapshotLoader[F[_]] {
+
+  def readGlobalSnapshot(hash: Hash): F[Either[Signed[GlobalSnapshot], Signed[GlobalIncrementalSnapshot]]]
+}
+
+object GlobalSnapshotLoader {
+
+  def make[F[_]: Async: KryoSerializer](
+    incrementalGlobalSnapshotLocalFileSystemStorage: SnapshotLocalFileSystemStorage[F, GlobalIncrementalSnapshot],
+    path: Path
+  ): F[GlobalSnapshotLoader[F]] =
+    SnapshotLocalFileSystemStorage.make[F, GlobalSnapshot](path).map(make(incrementalGlobalSnapshotLocalFileSystemStorage, _))
+
+  def make[F[_]: Async](
+    incrementalGlobalSnapshotLocalFileSystemStorage: SnapshotLocalFileSystemStorage[F, GlobalIncrementalSnapshot],
+    globalSnapshotLocalFileSystemStorage: SnapshotLocalFileSystemStorage[F, GlobalSnapshot]
+  ): GlobalSnapshotLoader[F] = new GlobalSnapshotLoader[F] {
+
+    def readGlobalSnapshot(hash: Hash): F[Either[Signed[GlobalSnapshot], Signed[GlobalIncrementalSnapshot]]] = {
```

Comment:
Knowing only the last incremental snapshot, I don't know what was before (full or incremental snapshot). 

> you iterate until the first ever incremental

I don't know what is the first ever incremental.

Reply:
>I don't know what is the first-ever incremental.

You could discriminate the first ever incremental by a value on some specific field, i.e. `SnasphotOrdinal` field, that would be hardcoded in the config related to the environment:

```
Testnet -> 123
Mainnet -> 456
Dev -> 1
```

Reply:
but I'd have to hardcode it and it won't be universal. that way it works well for each case and env

Reply:
Yes, but it doesn't have to be universal. We're doing the change from `Global` to `Incremental` only once in history (per environment).

Reply:
but what's the issue with reading "this or that"? it's also simple and don't require hardcoding or reading config and changing parameters. it's basically transparent for the end user to switch from global to incremental snapshots. and it's less error-prone, because you just read till first full snapshot.

Reply:
I'll take a look at that. 

Anyway, traversing forward only is not possible, as you'd have many possible paths and you may not finish with the same snapshot what is the current/latest one (see: forks).

Reply:
It is possible and simple. If my config says that the schema change is at the ordinal 123, then I'm loading: 123, 124, 125, ... until I find a snapshot with a hash matching the rollback hash (or hash & ordinal).

Reply:
assuming you have corresponding ordinals, which is bad assumption. but you can rely on hashes (but you don't know the next hash). As I said, you may not find the latest if you had forks or sth

Reply:
>Ordinal may point to a different snapshot (different hash) if you was in fork, but querying by hash, you always get what you asked for.

Not sure how this applies to the problem.

If you traverse forward by ordinal and validate history along the way and you eventually reach snapshot with hash equal to given rollback hash, you've loaded a valid history.

Reply:
The other possibility is that you could never create a chain (but traversing backward would create)

Reply:
What would be the case for not creating the chain?

Reply:
If next ordinal doesn’t match the majority hash (which I think is possible after fork). We can discuss it today. If we can assume we can ignore this or it won’t ever happen, then I’m good with it. Anyway, for chain correctness, I think we should traverse backward to find all the hashes and then forward to apply snapshots

Reply:
Traversing is implemented using fs2 solution you've provided. Can we move next changes (if needed) to the separate PR?

Reply:
Yes, I think we can leave it as it is for now.